src/client.rs

```diff
@@ -1,96 +1,126 @@
+    method: reqwest::Method,
     accepted: Option<HashSet<reqwest::StatusCode>>,
+}
+
+/// A link checker using an API token for Github links
+/// otherwise a normal HTTP client.
+#[derive(Builder, Debug)]
+#[builder(build_fn(skip))]
+#[builder(setter(into))]
+#[builder(name = "ClientBuilder")]
+pub struct ClientBuilderInternal {
+    github_token: Option<String>,
+    includes: Option<RegexSet>,
+    excludes: Excludes,
+    #[builder(default = "5")]
+    max_redirects: usize,
+    user_agent: String,
+    allow_insecure: bool,
+    scheme: Option<String>,
+    custom_headers: HeaderMap,
+    method: reqwest::Method,
+    accepted: Option<HashSet<http::StatusCode>>,
+    timeout: Option<Duration>,
     verbose: bool,
-    progress_bar: Option<&'a ProgressBar>,
 }
 
-impl<'a> Checker<'a> {
-    /// Creates a new link checker
-    // we should consider adding a config struct for this, so that the list
-    // of arguments is short
-    #[allow(clippy::too_many_arguments)]
-    pub fn try_new(
-        github_token: Option<String>,
-        includes: Option<RegexSet>,
-        excludes: Excludes,
-        max_redirects: usize,
-        user_agent: String,
-        allow_insecure: bool,
-        scheme: Option<String>,
-        custom_headers: HeaderMap,
-        method: RequestMethod,
-        accepted: Option<HashSet<http::StatusCode>>,
-        timeout: Option<Duration>,
-        verbose: bool,
-        progress_bar: Option<&'a ProgressBar>,
-    ) -> Result<Self> {
+impl ClientBuilder {
+    pub fn build(&mut self) -> Result<Client> {
         let mut headers = HeaderMap::new();
+
         // Faking the user agent is necessary for some websites, unfortunately.
         // Otherwise we get a 403 from the firewall (e.g. Sucuri/Cloudproxy on ldra.com).
+        // let user_agent = self.user_agent.as_ref().unwrap_or(&"lychee/0.3.0".to_string());
+        let user_agent = match self.user_agent {
+            Some(ref u) => u.clone(),
+            None => String::from("lychee/0.3.0"),
+        };
+
         headers.insert(header::USER_AGENT, HeaderValue::from_str(&user_agent)?);
         headers.insert(header::TRANSFER_ENCODING, HeaderValue::from_str("chunked")?);
-        headers.extend(custom_headers);
+        if let Some(custom) = &self.custom_headers {
+            headers.extend(custom.clone());
+        }
+
+        let allow_insecure = self.allow_insecure.unwrap_or(false);
+        let max_redirects = self.max_redirects.unwrap_or(DEFAULT_MAX_REDIRECTS);
 
         let builder = reqwest::ClientBuilder::new()
             .gzip(true)
             .default_headers(headers)
             .danger_accept_invalid_certs(allow_insecure)
             .redirect(reqwest::redirect::Policy::limited(max_redirects));
 
-        let builder = match timeout {
-            Some(timeout) => builder.timeout(timeout),
+        let builder = match self.timeout {
+            Some(t) => builder.timeout(t.ok_or_else(|| anyhow!("cannot read timeout"))?),
```

Comment:
I think it has more to do with `parsing` rather than `reading` the timeout value?